staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go

```diff
@@ -190,16 +190,29 @@ func (d *memCacheClient) refreshLocked() error {
 		return err
 	}
 
+	wg := &sync.WaitGroup{}
+	resultLock := &sync.Mutex{}
 	rl := map[string]*cacheEntry{}
 	for _, g := range gl.Groups {
 		for _, v := range g.Versions {
-			r, err := d.serverResourcesForGroupVersion(v.GroupVersion)
-			rl[v.GroupVersion] = &cacheEntry{r, err}
-			if err != nil {
-				utilruntime.HandleError(fmt.Errorf("couldn't get resource list for %v: %v", v.GroupVersion, err))
-			}
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				defer utilruntime.HandleCrash()
+
+				r, err := d.serverResourcesForGroupVersion(v.GroupVersion)
+
+				resultLock.Lock()
+				defer resultLock.Unlock()
+
+				rl[v.GroupVersion] = &cacheEntry{r, err}
+				if err != nil {
+					utilruntime.HandleError(fmt.Errorf("couldn't get resource list for %v: %v", v.GroupVersion, err))
```

Comment:
I'm assuming this is fast, but you're holding the lock while running this function.

Reply:
I've moved it up above the critical section. Thanks!